sdb/commands/linux/ps.py

```diff
@@ -0,0 +1,180 @@
+        # thread.
+        #
+        return shorten(s, width=50)
+    except drgn.FaultError:
+        #
+        # The command line information is contained in the user address
+        # space of each thread, rather than in the kernel's address
+        # space. Thus, often, it may not be possible to retreive the
+        # thread's command line; e.g. when reading from a core dump.
+        #
+        return ""
+
+
+class Ps(sdb.Locator, sdb.PrettyPrinter):
+    """
+    Locate and print information about processes
+
+    POTENTIAL COLUMNS
+        task - address of the task_struct
+        uid  - user id
+        pid  - the pid of the thread's process
+        time - cumulative CPU time, "[DD-]HH:MM:SS" format
+        state - the state of the thread
+        prio - the priority of the thread
+        comm - the thread's command
+        cmdline - the thread's command line (when available)
+
+    EXAMPLE
+        sdb> ps -e
+        TASK               UID  PID  TIME    STIME   PPID  CMD
+        ------------------ ---- ---- ------- ------- ----- ----------------
+        0xffff9995001a0000 1000 4387 0:00:00 0:00:00 1218  inotify_reader
+        0xffff9995001a2d00 1000 4382 0:00:00 0:00:00 1218  slack
+        0xffff9995001a4380 1000 4383 0:00:04 0:00:00 1218  slack
+        0xffff9995001a5a00 1000 4554 0:11:00 0:09:56 4118  AudioIP~ent RPC
+        0xffff99950a430000 1000 4284 0:00:00 0:00:00 4118  HTML5 Parser
+        0xffff99950a431680 1000 4157 0:00:12 0:00:08 1218  localStorage DBa
+        ...
+    """
+
+    names = ["ps"]
+    input_type = "struct task_struct *"
+    output_type = "struct task_struct *"
+
+    FIELDS: Dict[str, Callable[[drgn.Object], Union[str, int]]] = {
+        "UID":
+            lambda obj: int(obj.real_cred.uid.val),
+        "PID":
+            lambda obj: int(obj.pid),
+        "TIME":
+            lambda obj: str(
+                datetime.timedelta(seconds=int(obj.utime) / 1000 / 1000)),
+        "STIME":
+            lambda obj: str(
+                datetime.timedelta(seconds=int(obj.stime) / 1000 / 1000)),
+        "PPID":
+            lambda obj: int(obj.parent.pid),
+        "STAT":
+            lambda obj: str(Stacks.task_struct_get_state(obj)),
+        "CMD":
+            lambda obj: str(obj.comm.string_().decode("utf-8")),
+    }
+
+    @classmethod
+    def _init_parser(cls, name: str) -> argparse.ArgumentParser:
+        parser = super()._init_parser(name)
+        parser.add_argument('-e', '--every', action='store_true', \
+                help="Select all processes. Identical to -A")
+        parser.add_argument('-A', '--all', action='store_true', \
+                help="Select all processes. Identical to -e")
+        parser.add_argument('-C', '--C', type=str, \
+                help="Print only the process IDs of a command")
+        parser.add_argument('-x', '--x', action='store_true', \
+                help="Show PID, TIME, CMD")
+        parser.add_argument('-h', action='store_true', \
```

Comment:
Unfortunately `argparse` doesn't allow us to use `-h` for anything other than the help message. Fortunately for us looking at `man ps` I see that that to skip the headers only the long form argument exists `--no-headers` and `--no-heading`.

Reply:
updated.